src/backend/stream.rs

```diff
@@ -0,0 +1,821 @@
+                pa_stream_unref(self.output_stream);
+            }
+
+            if !self.input_stream.is_null() {
+                pa_stream_set_state_callback(self.input_stream, None, ptr::null_mut());
+                pa_stream_set_read_callback(self.input_stream, None, ptr::null_mut());
+                pa_stream_disconnect(self.input_stream);
+                pa_stream_unref(self.input_stream);
+            }
+            pa_threaded_mainloop_unlock(self.context.mainloop);
+        }
+    }
+
+    pub fn start(&mut self) -> i32
+    {
+        self.shutdown = false;
+        self.stream_cork(CorkState::uncork() | CorkState::notify());
+
+        if !self.output_stream.is_null() && self.input_stream.is_null() {
+            /* On output only case need to manually call user cb once in order to make
+             * things roll. This is done via a defer event in order to execute it
+             * from PA server thread. */
+            pa_threaded_mainloop_lock(self.context.mainloop);
+            pa_mainloop_api_once(pa_threaded_mainloop_get_api(self.context.mainloop),
+                                 Some(pulse_defer_event_cb),
+                                 self as *mut _ as *mut _);
+            pa_threaded_mainloop_unlock(self.context.mainloop);
+        }
+
+        cubeb::OK
+    }
+
+    pub fn stop(&mut self) -> i32
+    {
+        pa_threaded_mainloop_lock(self.context.mainloop);
+        self.shutdown = true;
+        // If draining is taking place wait to finish
+        while !self.drain_timer.is_null() {
+            pa_threaded_mainloop_wait(self.context.mainloop);
+        }
+        pa_threaded_mainloop_unlock(self.context.mainloop);
+
+        self.stream_cork(CorkState::cork() | CorkState::notify());
+
+        cubeb::OK
+    }
+
+    pub fn position(&self) -> Result<u64>
+    {
+        if self.output_stream.is_null() {
+            return Err(cubeb::ERROR);
+        }
+
+        let in_thread = pa_threaded_mainloop_in_thread(self.context.mainloop);
+
+        if in_thread == 0 {
+            pa_threaded_mainloop_lock(self.context.mainloop);
+        }
+
+        let mut r_usec: pa_usec_t = Default::default();
+        let r = pa_stream_get_time(self.output_stream, &mut r_usec);
+        if in_thread == 0 {
+            pa_threaded_mainloop_unlock(self.context.mainloop);
+        }
+
+        if r != 0 {
+            return Err(cubeb::ERROR);
+        }
+
+        let bytes = pa_usec_to_bytes(r_usec, &self.output_sample_spec);
+        let position = (bytes / pa_frame_size(&self.output_sample_spec)) as u64;
+        Ok(position)
+    }
+
+    pub fn latency(&self) -> Result<u32>
+    {
+        if self.output_stream.is_null() {
+            return Err(cubeb::ERROR);
+        }
+
+        let mut r_usec: pa_usec_t = 0;
+        let mut negative: i32 = 0;
+        let r = pa_stream_get_latency(self.output_stream, &mut r_usec, &mut negative);
+
+        debug_assert!(negative == 0);
+        if r != 0 {
```

Comment:
We shouldn't really assert the value of negative until after we've checked the return code, since the callee could write whatever it likes in there (I know this is wrong in the original code also)